src/main/java/uk/gov/hmcts/reform/draftstore/service/DraftService.java

```diff
@@ -0,0 +1,88 @@
+
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.draftstore.data.DraftStoreDAO;
+import uk.gov.hmcts.reform.draftstore.domain.CreateDraft;
+import uk.gov.hmcts.reform.draftstore.domain.Draft;
+import uk.gov.hmcts.reform.draftstore.domain.DraftList;
+import uk.gov.hmcts.reform.draftstore.domain.UpdateDraft;
+import uk.gov.hmcts.reform.draftstore.exception.AuthorizationException;
+import uk.gov.hmcts.reform.draftstore.exception.NoDraftFoundException;
+
+import java.util.List;
+import java.util.Objects;
+
+@Service
+public class DraftService {
+
+    private final DraftStoreDAO draftRepo;
+
+    public DraftService(DraftStoreDAO draftRepo) {
+        this.draftRepo = draftRepo;
+    }
+
+    public Draft read(String id, UserAndService userAndService) {
+        return draftRepo
+            .read(toInternalId(id))
+            .filter(draft -> Objects.equals(draft.userId, userAndService.userId))
+            .filter(draft -> Objects.equals(draft.service, userAndService.service))
+            .orElseThrow(() -> new NoDraftFoundException());
+    }
+
+    public DraftList read(UserAndService userAndService, Integer after, int limit) {
+        List<Draft> drafts =
+            draftRepo.readAll(
+                userAndService.userId,
+                userAndService.service,
+                after,
+                limit
+            );
+        return new DraftList(drafts);
+    }
+
+    public int create(CreateDraft newDraft, UserAndService userAndService) {
+        return draftRepo.insert(userAndService.userId, userAndService.service, newDraft);
+    }
+
+    public void update(String id, UpdateDraft updatedDraft, UserAndService userAndService) {
+        draftRepo
+            .read(toInternalId(id))
+            .map(d -> {
+                assertCanEdit(d, userAndService);
+                draftRepo.update(toInternalId(id), updatedDraft);
+                return d;
+            })
+            .orElseThrow(() -> new NoDraftFoundException());
+    }
+
+    public void delete(String id, UserAndService userAndService) {
+        draftRepo
+            .read(toInternalId(id))
+            .ifPresent(d -> {
+                assertCanEdit(d, userAndService);
+                draftRepo.delete(toInternalId(id));
+            });
+    }
+
+    private void assertCanEdit(Draft draft, UserAndService userAndService) {
+        boolean userOk = Objects.equals(draft.userId, userAndService.userId);
+        boolean serviceOk = Objects.equals(draft.service, userAndService.service);
+
+        if (!(userOk && serviceOk)) {
+            throw new AuthorizationException();
+        }
+    }
+
+    /**
+     * Converts external API id to internally used id.
+     */
+    private int toInternalId(String apiId) {
+        try {
+            // currently database ID is an int
+            return Integer.parseInt(apiId);
+        } catch (NumberFormatException exc) {
+            throw new NoDraftFoundException();
```

Comment:
`404` doesn't seem suitable for an unparsable parameter. I would suggest a `400` or `500` instead.

Reply:
I think 404 is the right code. The exception may be misleading, but what's really happening here is we're getting an ID of a non-existing draft.

Reply:
For me that's hiding the problem, making troubleshooting harder. Validation shouldn't let unparsable parameters through and reaching this code section should mean that there is a bug in the application.

Reply:
The background for this is that a requirement was added for the ID to be a string (on API level).  
@rlewan had a good point that we can keep it a string and add a regex (documented by Swagger).

Will investigate separately.

Reply:
I don't see how it makes it harder to troubleshoot. If the API provides a non-existing ID, it gets a 404 in every case. There's no reason to validate whether it's an integer or not. We don't require it to be an int and it's not an error in the service if someone provides something else.